src/userfunctions.jl

```diff
@@ -13,11 +13,23 @@ Return the current residual vector of the problem
 function getresidual end
 
 """
-    getjacobian(problem)
+    getsystemmatrix(problem,solver)
 
-Return the current jacobian of the problem
+Return the system matrix of the problem. For a Newton solver this
+method should return the Jacobian, while for a steepest descent method
+this can be a preconditioner as e.g., the L2 product of the gradients.
+By default the system matrix for the `SteepestDescent` method is the unity matrix
+and thus, renders a vanilla gradient descent solver.
 """
-function getjacobian end
+function getsystemmatrix end
+
+"""
+    getenergy(problem,𝐮)
```

Comment:
Would it make sense to either call this `calculate_energy(problem, u)` or redefine it to have
`getenergy(problem)` which relies on `update_problem(problem, du; jacobian=false, residual=false)` being called first? (In that case, care must be taken in `ArmijoGoldstein` L34 that a copy of the residual is taken in case the user decides that it isn't worth the effort to avoid recalculating the residual...)
Looking at the code the former seems to make most sense to me, but you know best for the line searches! In any case I think it makes sense to support the `jacobian::Bool` kwarg for gradient free methods, but it should be optional for the user to care (i.e. nothing bad should happen if it is updated).

Reply:
The current `ArmijoGoldstein` based rule is just backtracking, i.e. you look at the interval from one side, where it could be easily possible to rely on `update_problem`, but for Wolfe-Powell methods (I think) you search from two ends of the interval where it would be quite tricky to rely on update_problem, since you need to stash then different values instead of calling `calculate_energy` with a given `problem` and `u`. Nonetheless, it would be possible but could potentially complicate more sophisticated linesearches

The renaming is no problem for me, I can change that anyways

Reply:
Sounds good!